### ISSUE TITLE
Ideas, Feature Requests and Feature Outlining

### ISSUE THREAD

**Issue by DarkMio** [ValveARG/anticitizen-one#1]
(Not a native english speaker, so bare with me if my grammar goes off the rails)

Lets collect a bunch of ideas to properly plan a modular, extensible and performant software for ValveARG.

## List of questions:

Also we should answer a few questions beforehand:
- What worked well for ValveARG?
- What didn't work so well?
- What do we want to keep?
- How would we handle the next / future ARG?
- How can we reach the most ignorant people with a steady flow of information for a future ARG?

Next on the list would be:
- What features do you request, plan, sent in, sent back, queried, lost, found, subjected to public inquiry, lost again, and finally buried in soft peat for three months and recycled as firelighters about?
- How much time can you invest in general (just to not get cool ideas going and then die down)
- What kind of services are out there that we could simply incorporate instead of writing it again
- How would the new pages look like? Simply go with a framework and make some modifications or actually invest time in redesigning elements?

## List of ideas and suggestions:

- I'll take care about writing a one/two pager for a mailing/subscription list. Moddota.com (a side project of @penguinwizzard) has a mailer that can tank through >10k mails. On that list will be:
    - Message interface (notify, update, ...)
    - History page (last messages, sortable by timespan ...)
    - Update system (Send an update message and it relays that to twitter, IRC, Steam, Pushbullet, Whatsapp [that's a thing actually], telegram, ...)
    - API interface (in IRC is Rabscuttle, which is my bot, would be nice if you could notify 15k people by simply ``+notify Another ValveARG is going on, ...`` or something silly like this.
- Having Steam OAuth would be nice for all webservices. Incentives would be like:
    - Webchat autofills your credentials for you (yuck, that would mean storing IRC passwords in plaintext)
    - Mail System Admin Panel ...
    - Editor Features in Wikis (if we keep them and continue them)

**Comment by Penguinwizzard** [comment]
Just a note: having a sort of spec document would be good early on; I'm working on one for the ModDota site rewrite, and if it lines up enough with what's needed for the ValveARG site then we could potentially share development resources and end up with better services.

**Comment by xPaw** [comment]
Basically what worked well was having a wiki set up. However each ARG had it's own wiki, and that required manual setup from the guy that hosts valvearg.com. My initial idea was to have a single wiki with all args moved to it, then there could be shared pages, including pages with tools and stuff like that.

I'm guessing there should be a page in front of the wiki (as there is now) with some basic explanations, mailing list signup and stuff like that, because embedding that in the wiki itself probably wouldn't work so well.

**Comment by yaakov-h** [comment]
I'd think if we have a merged wiki, we should have a way to mark the sections for old ARGs as read-only, and have only current ARGs plus general pages as read-write.

**Comment by JakeRG** [comment]
More detailed comment will follow once I've had time to read the above, think about it, and write something more in-depth. For now, the main two things I want to point out are:

- I think we need a system through which we can actively gather everyone when a new ARG pops up, rather than passively wait until people find out about it through Twitter or articles on game websites. If we're there early, we can gather everyone to us, rather than having everyone spread out and organise their own groups. A mailing list is good, everything added to it is a nice bonus. Premade IFTTT triggers could enable less tech-savvy people to quickly get other methods.
- Secondly, previous ARGs (especially the Potato one due to its enormous scope) had issues with the stream of information. The wiki is good for documenting, but not for discussion. The IRC channels are good for discussion, but are quickly overwhelmed by tons of spam if everyone uses it at the same time. I believe we need an extra layer to tackle this issue - something to gather input, with a built-in system to give high-quality input high priority, while allowing comments and questions for expanding upon the input before discussing it in-depth and in real-time in IRC (followed by documenting it on the wiki if it's good). Something like... Reddit. Now, I'm not saying we need to use Reddit per se, but a voting / commenting system works really well to filter the good stuff from the crap, and could really help us battle the IRC spam we had during previous ARGs. Not having to give and take voice from IRC users while muting the channel for non-voiced users should free up a lot of time for the mods, too. Again, not sure if Reddit is our best option, but it might be tricky to set up something similar ourselves.

... Dammit I need to learn to make quick comments rather than write entire epics.

**Comment by DarkMio** [comment]
I already though about something like a timeline (which then needs branching) - so people can instantly see the current progress. Aomething like this - per ARG (taken from Amazon):

![img](http://i.imgur.com/ILdxkI3.png)

That could be probably pretty cool and should be doable with d3js. It's pretty powerful (Example: http://www.findtheconversation.com/concept-map#) and probably a git-like branch visualization would do it justice - maybe displayed on top oof the webchat client with a llittle flag saying "current Progress" or something silly like that.

**Comment by john-best** [comment]
IRC is definitely an issue when 3k+ members are on. The Steam Detectives discord group had a ton of problems with keeping users on-topic, even with 10-15 different channels. Of course, for IRC, kicking and banning are much easier and +m causes spam to quickly become a non-issue.

Reddit is an issue because people are going to upvote easily digested posts and comments (read: shitposts) because that's just how people are. Moderating on Reddit is going to be a pain in the ass with the current mod tools that they give us.

Not sure if we have any viable alternatives, though.

**Comment by DarkMio** [comment]
Stack Overflow Voting System is one, that is easily maintainable and able to cureate. I think reddit isn't the best plattform, since a few people should control the topic (for ARG progress) rather than anyone. A forum in general wouldn't be bad, but my opinion on forums is not the best - especially with many random people suddenly joining.

For visualizing, I still think that an automated branching-system with d3js would be something, that could get pretty and useful. It could even autoupdate by polling every ~5 minutes some dynamic data.

**Comment by JakeRG** [comment]
I agree that Reddit isn't the best platform, nor am I saying it's flawless. I am in fact not very active on it at all (I occasionally visit a link to it, and then write a comment once a year) because I'm not a great fan. However, I do not agree that "a few people should control the topic" - in my opinion, the whole idea of ARGs is to have tons of people come together to solve a mystery together. It is not up to us to decide who should control the topic. Besides, with a large enough ARG, it's hard to let a handful of people decide what is good information, and what isn't. 

Hence my suggestion of a Reddit-like voting and commenting system, where everyone can help out. IRC simply lacks the finesse required for filtering out the good from the bad quickly, and while +m makes spam a non-issue, it makes new people who found important info go elsewhere because they simply can't tell us anything. Yes, some shitposts will be upvoted. No, that doesn't mean the good stuff will not also rise to the top. People love it when there's an amazing new piece of shiny info, and they'll gladly upvote to feel like they were part of solving the puzzle. The same thing happened with the SteamSaleDetectives somewhat-ARG-like-thing, 

Obviously, we all realize that just adding an extra layer won't solve all our problems. Some issues for such a system that need addressing:
- Shitposts. We are not interested in Garfield Kart. Thus, we need a system that allows a handful of people to quickly determine what is or isn't bullshit, lock topics accordingly, and preferably (temp/perm) ban shitposters. 
- Related to the above: picking moderators. Usually, those who want to be one shouldn't be one, while those who are most capable aren't interested. We don't need trigger-happy mods that close topics for one joke in a comment, nor people that ban users on their first non-serious offence.
- Accounts. Allow everyone in, and you get the crapfest we had on the wiki before we disabled anonymous editing. Have users require a manually accepted account, and people might call you an elitist group of bastards, and regroup elsewhere. Which is bad. not because we're awesome and everyone should come to us (even though that is true), but because dividing the knowledge and brainpower slows us all down. Automatically accepted accounts are probably the best alternative, but it's only a small hurdle for wannabe spammers.
- Oh, and it should be capable of handling a few hundred people simultaneously. Not even counting bots and spam accounts.

Anyway, that's assuming people agree with me that such a system is useful. I'm saying it is, but do others feel we can manage just fine with what we had? Do you guys have any counter-arguments I didn't think of? Perhaps we can find or come up with something even better.

**Comment by DarkMio** [comment]
Alright, I think I have one potentional lead now:

# Node Graph over the chat window

Displaying a node graph on top of the chat can help us giving people an overview over the timeline, something like this:

![Git Graph](http://gugod.org/2009/12/12/3389620_29acb2fe86d3e03e7f8c665c4225c454.jpg)

We should probably chain a graph with either channel infos (this channel is working on that) and/or a forums subject on this. A reddit-like forum currently looks like a viable choice to get leads up and shit down. We need a more detailed look into it.

A D3JS graph for that is actually pretty easy. We could supply it with minute-based updates on it, it needs:

A view (iframe, div, whatever), a panel to edit a node and a link-method to either link people into channels, wiki pages or forum posts. Please ignore how we manage the edits, but I would suggest to not go overboard with new branches.

Coding that piece is actually easy. A parent node has (0, n) child nodes. A multiple parents can reference a single child. A database for that is easy, probably two tables that look something like that:

    ARG Table: ARG-ID | Name | Brief Description
    Node Table: Parent ID | Child-List | Parent-ID (reference) | Title | Description | Addition Data ...

Some examples: http://csclub.uwaterloo.ca/~n2iskand/?page_id=13 http://bl.ocks.org/Neilos/584b9a5d44d5fe00f779

It should be, in my imagination, a vertical directed graph that works as timeline. I can create a mockup when I am at home - the logic is pretty easy all in all. Technically it is also easy to use a REST API which we can cache pretty hefty - so load shouldn't be an issue either.

What would work with that is to be able to link to the unique-id of a graph (if possible) within IRC and it focuses that for a user that clicks it, that would make a conversation like this possible:

    RandomUser: Hey, have you seen that there is a link that accepts input?
    ARG Player: yes, we investigated this here: @aef178b3

Upon clicking it, it hovers over the node and displays full information on that information. This would make it easier to streamline informations and redirecting people to the right endpoint / disregarding infos that are days old.

What we still have to figure out is how we handle Wikis in the future and how we get stuff better coordinated. I can write some outlines for both of my planned systems (this node graph, the mailer) and plan out a modularized login system.

# Registering and Module System

I have to write some stuff down before I can tell you what a module system would need, but basically it should just be a rights management and user management system. It's easy to train Rabscuttle or Renol (IRC Bots) to let people chain their account together and automate some stuff with the bots.

The basic workflow would look like this: User signs in on ValveARG with their Steam OAuth. You config your account and can add a registered IRC account (type name in). Rabs or Renol privatemsg this registered account on IRC and asks him if his info is right (spam protection will be a problem probably). If he answers his message with "yes", his IRC account is associated with his ValveARG account.

Advantage would be, that he can now do several actions (based on our modules) from the bot - bot sends api request with a users registered username, which does whatever it should do. I have no concrete example which would need such chaining, besides some moderation tasks on our side.

# What we should discuss next:

- Module System. I really, really like the idea of having a framework where we can plug in some of our stuff and enable/disable them in the system.
- Handling user input. Uservoice, Reddit, Stackoverflow, Voting-Power Systems.
- Wiki Handling (imo we should have a centralized wiki for old args and a "working on" Wiki which we merge into the old-arg wiki once everything is done. That could, however, be rather tedious. Or we have a centralized Wiki where old stuff simply gets locked and visually embrace the current ongoing ARG (like taking 70-80% of the first front page).
- Design. Should we start to design our stuff on our own (pls no) or do we use a framework and modify it accordingly? Should we even start to make a media package (logos, corporate design, look and feel) to work on a framework. Which framework? My current suggestion is the Google Material Design - but that's my own preference, we should vote on that)
- Other modules (we probably don't need any more modules right now, maybe a user-handler to support wiki <-> valve arg users)

**Comment by xPaw** [comment]
> Wiki Handling (imo we should have a centralized wiki for old args and a "working on" Wiki which we merge into the old-arg wiki once everything is done. That could, however, be rather tedious. Or we have a centralized Wiki where old stuff simply gets locked and visually embrace the current ongoing ARG (like taking 70-80% of the first front page).

> Other modules (we probably don't need any more modules right now, maybe a user-handler to support wiki <-> valve arg users)

Merging old wikis and locking edits from public is fine and how we initially planned on doing it. On top of that, if the main site will have oauth login with Steam, that could be hooked up to handle wiki logins as well, so that if you login on the main site, you will be automatically logged in on the wiki.

> Design. Should we start to design our stuff on our own (pls no) or do we use a framework and modify it accordingly? Should we even start to make a media package (logos, corporate design, look and feel) to work on a framework. Which framework? My current suggestion is the Google Material Design - but that's my own preference, we should vote on that)

Just bootstrap it, later on can be modified to look nicer.

**Comment by DarkMio** [comment]
Muh bootstrap, but that's just my opinion. I can always invest a bit time inbetween for a nice logo (but I like the VA icon of this group very much).

We should lock out people of wiki edits (and registering) without being logged in on the ValveARG user handler. Just before we oversee it.

**Comment by xPaw** [comment]
> We should lock out people of wiki edits (and registering) without being logged in on the ValveARG user handler. Just before we oversee it.

Completely disable registration/login on wiki itself, and handle it on the main site.

https://www.mediawiki.org/wiki/AuthPlugin
https://www.mediawiki.org/wiki/Extension:ExtAuthDB

**Comment by DarkMio** [comment]
Right, that's exactly what I was thinking of. Would you use the Wiki as module or loosely chain the module system with the wiki authing?

I think having the wiki as module (unchanged, with some module infos) would make sense, but I trust your experience with that.

**Comment by xPaw** [comment]
> I think having the wiki as module (unchanged, with some module infos) would make sense, but I trust your experience with that.

I have zero experience in managing mediawiki, so no clue it would actually work :grin:

**Comment by DarkMio** [comment]
Heh, that I would plan it as module, something like:

- ValveARG
    - Core
    - Modules
        - Notifico
        - NodeGraph
        - MediaWiki

Each module supplies some infos. I just want to prevent table-spam and unify user control and management as seemless as possible. From there out on we could modify media wiki to query userdata from the core. This would also allow to disable the wiki as plugin or something like that.

Also thanks everyone, I like how this is going, this threads is very dense.

**Comment by JakeRG** [comment]
Regarding design, as I said in IRC: if there's a list of requirements, I can try to mock something up. My degree says I'm qualified for it, but I'm a bit rusty. =P It might take a bit depending on the requirements, but I'm guessing that's not an issue (we're not planning on finishing this in the next 2 weeks, are we?).

**Comment by TheScootz** [comment]
I have quite a bit of experiencing managing and using MediaWiki so if anywhere, that's where I would be most useful. #scootzforsysop

As far as discussion platforms, I'd like to rule out reddit and standard forums for the reasons already mentioned -- both are very easily polluted without heavy moderation. Whatever we decide to go with will need to be very autocratic if we're going to handle the volume of users we commonly experience during ARGs. Other people seem to have better ideas as to what would be best so I'll leave that up to them. :>

**Comment by TheCell** [comment]
### System between IRC and Wiki
I'm in favour of a **Stack Overflow system** instead of a Reddit based. But the rules on the actual Stack Overflow are way too hard.
#### The Good parts
* filter spam
* good commentors will probably be heared after a previous submission
* crowdsourcing ideas with occasional moderation
* semipermanent
#### The Bad parts
* it restricts comments too heavily (keeps comments on a minimum while maximizing the value/content). This is bad for one reason: A puzzle hunt needs a lot of discussion and suggestions leading to quantity of content and minimum of quality. The IRC is good to discuss and and suggest new things but gets out of hand too quickly. Things will be forgotten by the time they scroll past the top border

### Progress notes
A suggestion would be to have **(a) a Node Graph** over the chat window like @DarkMio suggested but I'd like to see a more **(b) static Graph** on the page. There should be some sort of **(b)** Timeline where you can see the actual discoveries lined up. That's good for keeping up with what happened the past x hours. After that you can glance at the **(a)** and see what we are up to at the mom

### Module System
I'd like to see a module system but best would be to have most of the stuff in the plugins. That way it would be easier for the Manager (Module System, can stay simpler) and easier for the plugin creator since you would not need a big list of "apis" and could develop it easier.

### Notifications
Only regarding Twitter here, lets pretend this is just a standin for all systems used to spread information. Maybe we could have 2 different Twitter accounts set up. One for big progress updates (as mentioned under Progress notes **(b) static Graph**) and one with frequent updates on what the community works on right now.

### General Sitetree
* Having a landing page with the chat and the  **(a) a Node Graph**
* Second page with **(b) static Graph** on top and the Stackoverflow Reddit malformation/hybrid
* Wiki with past ARG's in one place

### Ideas
Cool would be to have a get go page where known arg methods are listed. By that I mean a place with 
- links to different analyzers ( like [exif viewer](http://regex.info/exif.cgi) )
- a list with what we learned from past ARG's
- a list of known / used ARG hiding methods where you can go through and tick what you have tried

**Comment by xPaw** [comment]
I setup a very basic page: https://valvearg.github.io/
